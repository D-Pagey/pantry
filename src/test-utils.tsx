import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import React, { FC, ReactNode } from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { FirebaseContext } from './components/ProviderFirebase';

type AllTheProvidersProps = {
    children: ReactNode;
    customContext?: any;
};

const AllTheProviders: FC<AllTheProvidersProps> = ({ children, customContext }) => (
    <FirebaseContext.Provider
        value={{
            categories: [],
            deleteFoodItem: () => () => null,
            expiringCount: 0,
            isAuthed: false,
            isCheckingAuth: false,
            fridge: [],
            signOut: () => null,
            updateFridge: () => null,
            user: {
                email: null,
                name: null
            },
            ...customContext
        }}
    >
        <MemoryRouter>{children}</MemoryRouter>
    </FirebaseContext.Provider>
);

const customRender = (ui: any, customContext?: any, options?: any) =>
    render(ui, { wrapper: () => <AllTheProviders customContext={customContext}>{ui}</AllTheProviders>, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
