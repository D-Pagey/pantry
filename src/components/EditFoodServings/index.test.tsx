import React from 'react';
import userEvent from '@testing-library/user-event';
import { render } from '../../test-utils';
import { Fridge } from '../../fixtures';
import { EditFoodServings } from '.';

const props = {
    item: Fridge[0]
};

describe('EditFoodServings component', () => {
    it('should render', () => {
        const { container } = render(<EditFoodServings {...props} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should update remove correct servings amount', () => {
        const updateFridge = jest.fn();
        const { getByTestId } = render(<EditFoodServings {...props} />, { updateFridge });

        userEvent.click(getByTestId('2-0'));
        userEvent.click(getByTestId('3-0'));
        userEvent.click(getByTestId('EditFoodServingsSubmit'));

        expect(updateFridge).toHaveBeenCalledWith({
            batches: [
                { expires: expect.any(Date), owner: '1', servings: 1 },
                { expires: expect.any(Date), owner: '2', servings: 1 },
                { expires: expect.any(Date), owner: '3', servings: 3 }
            ],
            category: 'vegetables',
            name: 'carrots'
        });
    });

    it('should remove a batch if checked last serving', () => {
        const updateFridge = jest.fn();
        const { getByTestId } = render(<EditFoodServings {...props} />, { updateFridge });

        userEvent.click(getByTestId('1-0'));
        userEvent.click(getByTestId('EditFoodServingsSubmit'));

        expect(updateFridge).toHaveBeenCalledWith({
            batches: [
                { expires: expect.any(Date), owner: '2', servings: 2 },
                { expires: expect.any(Date), owner: '3', servings: 4 }
            ],
            category: 'vegetables',
            name: 'carrots'
        });
    });

    it('should update remove correct servings amount when checked and unchecked', () => {
        const updateFridge = jest.fn();
        const { getByTestId } = render(<EditFoodServings {...props} />, { updateFridge });

        userEvent.click(getByTestId('1-0'));
        userEvent.click(getByTestId('1-0'));

        userEvent.click(getByTestId('2-0'));
        userEvent.click(getByTestId('2-0'));

        userEvent.click(getByTestId('3-0'));
        userEvent.click(getByTestId('EditFoodServingsSubmit'));

        expect(updateFridge).toHaveBeenCalledWith({
            batches: [
                { expires: expect.any(Date), owner: '1', servings: 1 },
                { expires: expect.any(Date), owner: '2', servings: 2 },
                { expires: expect.any(Date), owner: '3', servings: 3 }
            ],
            category: 'vegetables',
            name: 'carrots'
        });
    });
});
