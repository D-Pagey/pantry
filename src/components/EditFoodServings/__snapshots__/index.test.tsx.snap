// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditFoodServings component should render 1`] = `
.c2 {
  background-color: #348841;
  border: 0;
  border-radius: 5px;
  color: white;
  cursor: pointer;
  padding: 1rem;
  min-width: 68px;
}

.c3 {
  background-color: white;
  border: 1px solid #348841;
  color: #348841;
}

.c0 {
  justify-self: center;
  margin: 0;
  padding: 0;
  list-style: none;
  width: 17rem;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 1rem 0;
}

.c5 {
  color: #EB5757;
  margin: 0 0 0 1rem;
}

.c8 {
  color: #FCBA3B;
  margin: 0 0 0 1rem;
}

.c10 {
  color: #348841;
  margin: 0 0 0 1rem;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: none;
  border: 1px solid darkgrey;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 1rem;
  border-radius: 50%;
}

.c4 {
  border: 1px solid #EB5757;
}

.c7 {
  border: 1px solid #FCBA3B;
}

.c9 {
  border: 1px solid #348841;
}

<ul
  class="c0"
>
  <li
    class="c1"
  >
    <button
      class="c2 c3 c4"
      type="button"
    >
      <span
        class="c5"
      >
        Expires in 
        0 seconds
      </span>
    </button>
    <button
      class="c6"
      data-testid="deleteServing"
      type="button"
    >
      <img
        alt="delete"
        src="delete.svg"
      />
    </button>
  </li>
  <li
    class="c1"
  >
    <button
      class="c2 c3 c7"
      type="button"
    >
      <span
        class="c8"
      >
        Expires in 
        2 days
      </span>
    </button>
    <button
      class="c6"
      data-testid="deleteServing"
      type="button"
    >
      <img
        alt="delete"
        src="delete.svg"
      />
    </button>
  </li>
  <li
    class="c1"
  >
    <button
      class="c2 c3 c7"
      type="button"
    >
      <span
        class="c8"
      >
        Expires in 
        2 days
      </span>
    </button>
    <button
      class="c6"
      data-testid="deleteServing"
      type="button"
    >
      <img
        alt="delete"
        src="delete.svg"
      />
    </button>
  </li>
  <li
    class="c1"
  >
    <button
      class="c2 c3 c9"
      type="button"
    >
      <span
        class="c10"
      >
        Expires in 
        5 days
      </span>
    </button>
    <button
      class="c6"
      data-testid="deleteServing"
      type="button"
    >
      <img
        alt="delete"
        src="delete.svg"
      />
    </button>
  </li>
  <li
    class="c1"
  >
    <button
      class="c2 c3 c9"
      type="button"
    >
      <span
        class="c10"
      >
        Expires in 
        5 days
      </span>
    </button>
    <button
      class="c6"
      data-testid="deleteServing"
      type="button"
    >
      <img
        alt="delete"
        src="delete.svg"
      />
    </button>
  </li>
  <li
    class="c1"
  >
    <button
      class="c2 c3 c9"
      type="button"
    >
      <span
        class="c10"
      >
        Expires in 
        5 days
      </span>
    </button>
    <button
      class="c6"
      data-testid="deleteServing"
      type="button"
    >
      <img
        alt="delete"
        src="delete.svg"
      />
    </button>
  </li>
  <li
    class="c1"
  >
    <button
      class="c2 c3 c9"
      type="button"
    >
      <span
        class="c10"
      >
        Expires in 
        5 days
      </span>
    </button>
    <button
      class="c6"
      data-testid="deleteServing"
      type="button"
    >
      <img
        alt="delete"
        src="delete.svg"
      />
    </button>
  </li>
</ul>
`;
